First stab at making Mudlet's text window readable by screenreaders

bauermann opened the pull request:
Hello! This is what I have so far regarding fixing issue #3342. It’s not done yet because there’s a significant issue I’m still trying to figure out how to fix (ideas welcome):

When new text arrives in the TTextEdit, the screen reader doesn’t announce it even though the widget is sending the ‘TextInsertEvent’ with the new text. The only time it works is when Caret Mode is enabled (press F7 to activate it), so what I believe is going on is that the screen reader only pays attention to the emitted ‘TextInsertEvent’ when the TTextEdit is focused.

Therefore I’m trying to figure out a way to tell the screen reader that changes in TTextEdit should be announced even when it’s not focused. Orca’s documentation mentions something called “live regions” which seems to be what I need, but so far I have the impression that it’s a concept applied only to web content. I hope I’m wrong though.

Using Orca’s “Speak entire document” command (KP_PLUS or <Insert> + Semicolon) works though, and so does speaking when moving the cursor in Caret Mode which means that the widget’s contents are being exposed correctly to the screen reader. It’s just that there’s something more I need to do in order to make the screen reader know that it needs to announce new text as it arrives.

Nobody asked to implement Caret Mode but I did anyway, inspired by the feature of the same name in Firefox. It seems useful to me, please let me know what you think.

I dropped the commit where I was removing the HTML markup from the tooltips in TConsole’s buttons. The screen reader still speaks them so it’s a bit annoying, but now there’s only `<p>…</p>` markup whereas before it was `<html><head/><body><p>…</p></body></html>`, so a lot better now.

Also, since Qt does expect HTML markup in the tooltip then it should strip the
markup before passing the tooltip contents along to the screen reader. So it’s a
problem to be fixed somewhere else in the technology stack.

I’m testing with GNOME on Linux, with the Orca screen reader.

I’d appreciate more testing, as well as feedback on the code and design.

I don’t expect this PR to be merged. It’s mostly so that the CI can generate snapshot builds.

vadi2 committed:
Add in files

vadi2 committed:
Wire it in

bauermann committed:
.gitignore: Ignore ‘compile_commands.json’

bauermann committed:
TAccessibleConsole: Move factory installation to main()

bauermann committed:
TAccessibleTextEdit: Don't override child methods

Rely on the superclass' implementation.

bauermann committed:
TAccessibleTextEdit: Add 'override' keyword to state()

bauermann committed:
TAccessibleTextEdit: Remove role() override

It's redundant, because QAccessibleWidget's constructor already takes a
role argument.

bauermann committed:
TAccessibleTextEdit: Add 'marqueed' to state() result

bauermann committed:
TAccessibleTextEdit: Use correct class name in debug output

bauermann committed:
TAccessibleTextEdit: Implement selection methods

Support only selection 0.

bauermann committed:
Minor improvements to debug messages

Use plain strings rather than QStringLiterals. The QStringLiteral()
versions add annoying double quotes to the output.

Also, use function calls rather than streams to output the messages. The
syntax is more compact and allows for better control of the formatting.

bauermann committed:
TAccessibleTextEdit::text(): Return contents in the requested range

The current cursor position isn't relevant in the calculation.

bauermann committed:
TAccessibleTextEdit: Implement textBeforeOffset(), textAfterOffset() and textAtOffset()

bauermann committed:
Move most of the code from TAccessibleConsole to TAccessibleTextEdit

TConsole is best viewer as a container of other widgets, rather than a
widget itself. Therefore, it makes more sense to add accessibility support
to its inner widgets.

bauermann committed:
TAccessibleTextEdit: Convert offset to linear position within actual contents

... as oposed to assuming that every (x, y) coordinate contains a character.

bauermann committed:
TAccessibleTextEdit: Change visibility of some methods to 'private'

This communicates that they're internal helpers for the class implementation and
aren't used anywhere else.

bauermann committed:
TAccessibleTextEdit: Validate offset parameters

bauermann committed:
TAccessibleTextEdit: No need to explicitly support ActionInterface in interface_cast()

QAccessibleWidget's version already does that for us.

Also take the opportunity to remove some unnecessary headers.

bauermann committed:
Implement caret mode

bauermann committed:
TAccessibleTextEdit: Implement text(QAccessible::Text t) method

This fixes the following warning:

  'TAccessibleTextEdit::text' hides overloaded virtual function

Adjust lineForOffset() and cursorPosition() to take into account the
newlines added by join().

Also, simplify text(int startOffset, int endOffset) using the new
text(QAccessible::Text t) method.

bauermann committed:
TAccessibleTextEdit: Consider offset at the end of the contents to be on the last char

bauermann committed:
TTextEdit: Small clarity improvements to drawForeground()

First, rename mLastRenderBottom to mLastRenderedOffset. The old name is
misleading because it actually holds the line that was at the top of the
screen when it was last rendered, not the bottom.

Also add explanation of what mScrollVector means.

These changes help understand the logic at TTextEdit::drawForeground().

bauermann committed:
TAccessibleTextEdit: Implement the attributes() method

It's partially based on QAccessibleTextWidget::attributes() from Qt 5.15.

bauermann committed:
TAccessibleTextEdit: Take scrolling into account in characterRect()

bauermann committed:
TAccessibleTextEdit: Implement method offsetAtPoint()

bauermann committed:
TAccessibleTextEdit: Implement method scrollToSubstring()

bauermann committed:
TConsole: Set accessibleName for mUpperPane and mLowerPane

bauermann committed:
TTextEdit: Use more specific ‘TextInsertEvent’ when new lines are added

Also, remove ‘DocumentContentChanged’ event from TConsole::print() methods
because both call TTextEdit::showNewLines(), which already sends
‘TextInsertEvent’.

bauermann committed:
TAccessibleConsole: Remove qDebug() statements

vadi2 commented:
Nice! Thanks for this. Will review it in detail in the coming days. In the meantime I posted it on our Discord's accessibility channel for people to start testing.

bauermann commented:
Em quarta-feira, 23 de fevereiro de 2022, às 17:22:43 -03, Vadim Peretokin 
escreveu:
> Nice! Thanks for this. Will review it in detail in the coming days. In
> the meantime I posted it on our Discord's accessibility channel for
> people to start testing.

That’s great! Thank you.

SlySven commented on src/TAccessibleTextEdit.h:
```diff
@@ -0,0 +1,97 @@
+
+    static QAccessibleInterface* textEditFactory(const QString &classname, QObject *object)
+    {
+        QAccessibleInterface *interface = 0;
```
Use `nullptr` rather than `0`please.

bauermann replied:
You're right. Done.

SlySven commented on src/TAccessibleTextEdit.h:
```diff
@@ -0,0 +1,97 @@
+/***************************************************************************
+ *   Copyright (C) 2008-2013 by Heiko Koehn - KoehnHeiko@googlemail.com    *
+ *   Copyright (C) 2014-2017 by Ahmed Charles - acharles@outlook.com       *
+ *   Copyright (C) 2014-2020 by Stephen Lyons - slysven@virginmedia.com    *
```
Don't forget your own Copyright line in all the files you modify please...

bauermann replied:
Oh, right! I forgot. Thank you for reminding me.

SlySven commented on src/CMakeLists.txt:
```diff
@@ -144,6 +144,7 @@ set(mudlet_SRCS
     TTabBar.cpp
     TTextCodec.cpp
     TTextEdit.cpp
+    TAccessibleTextEdit.cpp
```
Can we add the new class files in (case-insensitive) alphabetical order please - it makes it simpler to check that they are all there..! :grinning:

bauermann replied:
Sure! I thought it would make sense to add the accessibility classes next to the classes they pair with, but alphabetical is indeed simpler. Done.

SlySven commented on src/TConsole.cpp:
```diff
@@ -93,6 +97,17 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
+            } else if (mType & SubConsole) {
+                setAccessibleName(tr("Sub Console"));
+                setAccessibleDescription(tr("Sub console messages are shown here."));
+            } else if (mType & UserWindow) {
```
Technically you only need two `if`/`else if` in this set as an `else` will handle the third type that the `if` around them all will let through...

bauermann replied:
That's true. Changed.

SlySven commented on src/TConsole.cpp:
```diff
@@ -103,6 +118,9 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
             mCommandFgColor = mpHost->mCommandFgColor;
+
+            setAccessibleName(tr("Main Window"));
+            setAccessibleDescription(tr("Game messages are shown here"));
```
Maybe you need to mention that it can also contain sub-consoles (and a mapper) window? Also, is "Game messages" the best description for the content of this window?

bauermann replied:
Ah, I wasn’t aware that there could be sub-consoles. Is the following better?

`setAccessibleDescription(tr("Game content is shown here. It may contain subconsoles and a mapper window."));`

vadi2 replied:
Not sure that will be helpful to repeat this very often, because this text will be read very often. Let's wait until we have feedback that this is necessary before adding something potentially unnecessary.

SlySven commented on src/TConsole.cpp:
```diff
@@ -103,6 +118,9 @@ TConsole::TConsole(Host* pH, ConsoleType type, QWidget* parent)
             mCommandFgColor = mpHost->mCommandFgColor;
+
+            setAccessibleName(tr("Main Window"));
+            setAccessibleDescription(tr("Game messages are shown here"));
```
Maybe you need to mention that it can also contain sub-consoles (and a mapper) window? Also, is "Game messages" the best description for the content of this window?

bauermann replied:
I’ll mark this one as resolved since it seems to be a duplicate of the comment above. Feel free to reopen if that’s not the case.

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -321,6 +330,28 @@ void TTextEdit::showNewLines()
+
+
+    if (QAccessible::isActive()) {
+        QString newLines = QString();
```
It is unnecessary to assign a null `QString` the default constructor does that anyhow...
```suggestion
        QString newLines;
```

bauermann replied:
Ah, nice to know! Thank you for the suggestion.

I modified the original commit making this change, so I didn’t use the “Commit suggestion” GitHub feature which would add a commit with your name to the branch.

If you prefer me to use the “Commit suggestion” feature, just let me know and I’ll be glad to do it.

SlySven replied:
No, you don't have to use "Commit suggestion" - it is sometimes convenient but not always...

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -466,6 +497,13 @@ void TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen, in
     }
+
+    // If caret mode is enabled and the line is empty, still draw the caret.
+    if (mudlet::self()->isCaretModeEnabled() && mCaretLine == lineNumber && lineText.size() == 0) {
```
ℹ️ We generally use `isEmpty()` (Qt things)/ `empty()` (std things) rather than a `size()` or `count() == 0` as it gives a clearer signal as to the intent of the code...

bauermann replied:
I agree it’s clearer. I made the change here and also in TTextEdit::updateCaret() and TTextEdit::keyPressEvent().

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -466,6 +497,13 @@ void TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen, in
+
+    // If caret mode is enabled and the line is empty, still draw the caret.
+    if (mudlet::self()->isCaretModeEnabled() && mCaretLine == lineNumber && lineText.size() == 0) {
+        int charWidth = getGraphemeWidth(getGraphemeBaseCharacter("a"));
```
Why an `a` ? I'd have to check but isn't this always going to be `1` - or have I mistaken it for the width in pixels not the width in "normal" width characters... 😕

bauermann replied:
Hm, good question. I don’t remember anymore. I think I did that because I thought that perhaps the width of `a` depends on the locale. For example, in Chinese/Japanese most graphemes have width 2, so perhaps they use width 2 for Latin characters as well?

I don’t know if that is actually the case though. I will try to find out.

SlySven replied:
> ...For example, in Chinese/Japanese most graphemes have width 2...

Yeah the ones with a **W**ide width - which in those locales is pretty much all of them - and some codepoints are the wide equivalents of an unspecified corresponding narrow (normal) width ones - yet some other East-Asion codepoints have an "Ambiguous" width!

bauermann replied:
I tested with the zh_TW locale and even then `charWidth` is 1, so I'm removing this call to `getGraphemeWidth()`. Thank you for noticing this unnecessary piece of code.

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -602,7 +640,8 @@ int TTextEdit::drawGraphemeBackground(QPainter& painter, QVector<QColor>& fgColo
     QColor bgColor;
-    if (Q_UNLIKELY(static_cast<bool>(attributes & TChar::Reverse) != charStyle.isSelected())) {
+    bool caretIsHere = mudlet::self()->isCaretModeEnabled() && mCaretLine == line && mCaretColumn == column;
+    if (Q_UNLIKELY(static_cast<bool>(attributes & TChar::Reverse) != (charStyle.isSelected() != caretIsHere))) {
```
I'm not sure of this - originally it was to EXCLUSIVE-OR the reverse text mode with the reverse effect that selecting text does (so when combined the state is the same as if neither was applied) - what does the "CaratMod" mode do to how the text is/should be shown?

bauermann replied:
This is to implement the caret mode’s cursor. It is displayed as a reversed-color character cell, similar to old terminals.

The change above reverses the color when the character cell has normal color, and uses normal color when it has reversed-color.

SlySven replied:
So logically you have to handle all three conditions at the same time and apply the reverse effect when an ***odd*** number of them are true. Maybe the logic you have chosen will work, I'll have to think about it a bit more.

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -2581,3 +2620,128 @@ void TTextEdit::reportCodepointErrors()
+void TTextEdit::updateCaret()
+{
+    int lineOffset = imageTopLine();
+    if (!mIsLowerPane && mCaretLine < lineOffset) {
+        scrollTo(mCaretLine);
+    } else if (!mIsLowerPane && mCaretLine >= lineOffset + mScreenHeight) {
+        int emptyLastLine = mpBuffer->lineBuffer.last().length() == 0;
+        if (mCaretLine == mpBuffer->lineBuffer.length() - 1 - emptyLastLine) {
+            scrollTo(mCaretLine + 2);
+        } else {
+            scrollTo(mCaretLine + 1);
+        }
+    }
```
Refactoring out the `!mIsLowerPane` common term:
```suggestion
    if (!mIsLowerPane) {
        if (mCaretLine < lineOffset) {
            scrollTo(mCaretLine);
        } else if (mCaretLine >= lineOffset + mScreenHeight) {
            int emptyLastLine = mpBuffer->lineBuffer.last().length() == 0;
            if (mCaretLine == mpBuffer->lineBuffer.length() - 1 - emptyLastLine) {
                scrollTo(mCaretLine + 2);
            } else {
                scrollTo(mCaretLine + 1);
            }
        }
    }
```

bauermann replied:
Great! This indeed looks better.

Again, I made this change directly in the commit that added this code. Please let me know if you would prefer me to use the “Commit suggestion” feature. It’s just that I’m used to the workflow of improving the individual commits in a PR branch until they are “perfect”.

Just keep adding commits on top of the branch feels to me like working on a Jenga tower. :)

SlySven replied:
Yeah, but they will all go away when it gets **Squashed-and-merged** (our preferred choice) when the PR is accepted and merged into the `development` branch - hopefully the core-team member will take the time to sensibly combine all the commit messages into a single sensible one at the time.

SlySven commented on src/main.cpp:
```diff
@@ -167,6 +169,9 @@ int main(int argc, char* argv[])
     QApplication::setAttribute(Qt::AA_ShareOpenGLContexts);
 #endif
 
+    QAccessible::installFactory(TAccessibleConsole::consoleFactory);
```
This is being done before the `QApplication` instance is created - is this intended/required? {Some thing need to be done like this and some must not - but I do not know which applies for this...}

bauermann replied:
Good question. I couldn’t find anything saying whether it’s one or the other.
The only thing I found was this example, where it is done after the `QApplication` instance is created:

https://doc.qt.io/qt-5/accessible-qwidget.html#implementing-interface-factories

So I changed the `installFactory()` calls to right after the `QApplication` instantiation.

SlySven commented on src/mudlet.cpp:
```diff
@@ -2976,6 +2977,12 @@ void mudlet::slot_multi_view(const bool state)
 
+void mudlet::slot_caret_mode(bool state)
+{
+    // FIXME: Support multiple hosts.
```
Yeah, you will need to handle multiple `host` instances - and using `// FIXME:` is not making this PR "dangerous" (upsetting the *Danger* Bot that watches some aspects of PRs!) whereas using `// TODO:` definitely would... 😲

bauermann replied:
I plan to implement this before making the definitive PR. This comment is just so that I don’t forget about it. :)

SlySven reviewed (commented):
Just an initial review...!

Hadi-1624 commented:
I believe this is not working in windows at all.
Testing with NVDA screen reader on windows 10. when mudlet connects to a mud, the keyboard focus is on the input field and will refuse to  go anywhere else, even if I press the f7 key.

Just to make sure, here's how I imagine the client could be accessible:
1. screen reader reads incoming text in the output window
2. Mudlet by default has the keyboard focus on the input window. there should be an implementation in which the keyboard focus and cursor moves from the input to the output window, so that the output window could be navigated by the arrow keys and be read. this preferably should be switched by a hotkey so  it could be fast moving between the input and output windows.
3. Functions that sends the last 10 lines in the output window to the screen reader, which then could be assigned to hotkeys, so pressing control+shift+5 for example, would send the fifth line from bottom  to the scren reader to be read.

ironcross32 commented:
Hi, the F7 mode isn't activating in this case. I can verify that a connection has been made to the MUD I intended by running OCR on the game window. I am testing on Windows 10.

bauermann commented:
Hello Hadi,

Thank you very much for testing! It’s very helpful. I’m sorry that you had 
such a bad experience. I thought that there could be issues on Windows, 
but I had hope it wouldn’t be as bad as it it. :(

Em sexta-feira, 25 de fevereiro de 2022, às 07:01:15 -03, Hadi Rezaei 
escreveu:
> I believe this is not working in windows at all.

Ouch. I’m sorry about that. I don’t have access to a Windows machine, but I 
will do some research and see if I can find something out. Perhaps even try 
it on Wine (a Windows emulator for Unix).

> Testing with NVDA screen reader on windows 10. when mudlet connects to a
> mud, the keyboard focus is on the input field and will refuse to  go
> anywhere else, even if I press the f7 key.

I’m surprised that F7 isn’t working. I’m just using a conventional Qt API 
to add that hot key. I’ll dig into it.

It’s also possible to enable caret mode using the menu item “Caret Mode” 
under the ”Options” menu. It should have exactly the same effect as the F7 
key. Would you mind checking whether it works that way?

> Just to make sure, here's how I imagine the client could be accessible:

Thank you for this detailed explanation. It’s very helpful.

> 1. screen reader reads incoming text in the output window

I agree. This isn’t working even on my machine and it’s the most glaring 
and embarassing problem with my branch.

> 2. Mudlet by default has the keyboard focus on the input window. there
> should be an implementation in which the keyboard focus and cursor moves
> from the input to the output window, so that the output window could be
> navigated by the arrow keys and be read. this preferably should be
> switched by a hotkey so  it could be fast moving between the input and
> output windows.

This is what I’m implementing with Caret Mode. It’s unfortunate that it 
isn’t working on Windows. I’ll try to fix it.

> 3. Functions that sends the last 10 lines in the output
> window to the screen reader, which then could be assigned to hotkeys, so
> pressing control+shift+5 for example, would send the fifth line from
> bottom  to the scren reader to be read.

This is a great idea! Shouldn’t be hard to implement (famous last words). 
I’ll work on it.

bauermann commented:
Hello ironcross32,

Thank you for testing! I’m sorry that my branch failed so miserably. I was 
expecting problems, but thought it wouldn’t be this bad.

Em sexta-feira, 25 de fevereiro de 2022, às 13:22:38 -03, ironcross32 
escreveu:
> Hi, the F7 mode isn't activating in this case. I can verify that a
> connection has been made to the MUD I intended by running OCR on the
> game window. I am testing on Windows 10.

I’m sorry about that. I’m really suprised by this problem, since I 
implemented the F7 hot key using a standard Qt API...

It’s also possible to enable caret mode using the menu item “Caret Mode” 
under the ”Options” menu. It should have exactly the same effect as the F7 
key. Would you mind checking whether it works that way?

MuhammadHassan0000 commented:
hi,
i totally agree with what was said, however let me clarify something
what they ment, a window which contains all text received by the mud, and input window contains all the  commands i send
making it easier for us that way, both are separated and can be reached via a hot key
i hope this gets implimented
thanks
mohammed
can't wait to see this huge project coming out

ironcross32 commented:
I've posted some messages in the Mudlet Discord under the Mudlet development channel that may be of assistance.

bauermann committed:
Use ‘nullptr’ to initialize pointer in TAccessibleConsole and TAccessibleTextEdit

Thanks to Stephen Lyons for the suggestion.

bauermann committed:
TConsole: Improve accessible description of MainWindow consoles

Thanks to Stephen Lyons for the suggestion.

bauermann committed:
TConsole: Simplify if/else if statements

Thanks to Stephen Lyons for the suggestion.

bauermann commented:
Hello Mohammed,

Em sábado, 26 de fevereiro de 2022, às 17:22:22 -03, mohammedradwan2003 
escreveu:
> i totally agree with what was said, however let me clarify something
> what they ment, a window which contains all text received by the mud, and
> input window contains all the  commands i send making it easier for us
> that way, both are separated and can be reached via a hot key i hope
> this gets implimented

Thank you for explaining your idea. So you mean that Mudlet should have an 
accessibility mode where there are two separate windows, one for input and 
one for output?

That is is interesting, though if I understand correctly it would be out of 
scope for this specific issue (unless someone from the Mudlet project 
corrects me and says that it is in scope) so I cannot say that I would be 
able to implement it.

I think it’s worth opening a new GitHub issue about it so that there can be 
a discussion focused specifically on your idea.

> can't wait to see this huge project coming out

Me too! And it’s been a larger project than I originally thought it would 
be. :)

bauermann commented:
Em sábado, 26 de fevereiro de 2022, às 17:40:12 -03, ironcross32 escreveu:
> I've posted some messages in the Mudlet Discord under the Mudlet
> development channel that may be of assistance.

Thank you, I just saw it.

Hadi-1624 commented:
Hi there bauermann
Thanks for working on accessibility of the client!

I tried to follow your suggestion and enable Caret mode from the options menu but i've faced another problem where, after connecting to a game, pressing alt will not bring the menus up. i've tried  pressing alt, pressing it again and pressing down arrow but no luck.
Is there a way to open the option menu by a hotkey?
I do confirm that mudlet default hotkeys work however. package manager and module manager for example, can be opnned by their shortcuts.

bauermann commented:
Hello Hadi,

Em sábado, 26 de fevereiro de 2022, às 19:09:15 -03, Hadi Rezaei escreveu:
> Thanks for working on accessibility of the client!

You’re welcome! I hope I can bring it to a usable state.

> I tried to follow your suggestion and enable Caret mode from the options
> menu but i've faced another problem where, after connecting to a game,
> pressing alt will not bring the menus up. i've tried  pressing alt,
> pressing it again and pressing down arrow but no luck. Is there a way to
> open the option menu by a hotkey?
> I do confirm that mudlet default hotkeys work however. package manager
> and module manager for example, can be opnned by their shortcuts.

I have the same results in Linux. I was also not able to access the menu 
just using keyboard input, and had the same behavior as you describe 
(including being able to bringing up the package manager).

I’ll look into this issue as well.

bauermann commented:
Here are ironcross32’s comments on Discord. They allowed me to post here:

> Currently, I am not able to test the latest PR because it seems to be for
> Linux users only and I am on Windows. But reading the synopsis of the
> PR, I have a few concerns.
> The first of which is that while a Firefox style caret browsing mode
> sounds great in theory, it would have to stop the output from scrolling
> while it is active.

Indeed. And it does  stop scrolling. The good news is that when caret mode 
is enabled, the output widget has focus and thus the screen reader reads 
the new text as it arrives at the bottom.

> Second is the keystroke used to access it takes one
> hand off the keyboard long enough to tap it. something like tab or
> control + tab would work better in my opinion.

Good point. tab is already used for completion, but I think I can change 
the shortcut to control+tab. I will do that.

> The third thing is that
> to make the screen reader see all incoming text could, in theory, be
> easily done by using something like https://github.com/dkager/tolk.
> Though I believe this would only work under Windows, unfortunately.

Indeed, unfortunately tolk is a Windows-only solution from what I can see.
Theoretically, Qt should be able to provide cross-platform screen reader 
functionality though, and tolk or something like it shouldn’t be necessary. 
I must be doing something wrong in my code.

> It seems as if from the little bit of digging around that I've done, not
> actually being able to use Mudlet, that it supports plugins or packages.
> So, if a plugin implemented a catch-all trigger which sent all of the
> text contained within that trigger to the screen reader using Tolk, that
> would be sufficient for us to start using it.

IMHO it’s too early to give up on the current approach, as Qt theoretically 
should work. If all else fails though, then perhaps this could be 
considered (though I wouldn’t be the one implementing it since I don’t 
develop on Windows).

> As a P.S. the author of the PR was correct that live regions are for web
> content, and wouldn't be useful in making the screen reader detect
> incoming text.

Thank you! This is very useful information. If someone knows of a feature 
equivalent to live regions but for desktop apps, I’d love to know about it.

ironcross32 commented:
I'm 85% sure that QT will not do what you want. QT does have an accessibility layer, but it is for the controls. All of the controls unrelated to the actual game window do already seem to work. What needs to happen here is that text needs to be captured and sent to the running screen reader through its API. Unfortunately, there is no solution for this that fits all applications due to working cross-platform, and every screen reader has its own API. So there would need to be contingencies for sending through speech dispatcher in Linux and either writing your own class or using Tolk or Universal Speech on Windows. I have no clue about Mac.

This should be the first priority, because anything else is unimportant until incoming text is automatically read. Maybe if you can get this working under Linux, someone else can come along and duplicate it for Windows.

bauermann commented:
Em sábado, 26 de fevereiro de 2022, às 22:31:58 -03, ironcross32 escreveu:
> I'm 85% sure that QT will not do what you want. QT does have an
> accessibility layer, but it is for the controls. All of the controls
> unrelated to the actual game window do already seem to work. What needs
> to happen here is that text needs to be captured and sent to the running
> screen reader through its API. Unfortunately, there is no solution for
> this that fits all applications due to working cross-platform, and every
> screen reader has its own API. So there would need to be contingencies
> for sending through speech dispatcher in Linux and either writing your
> own class or using Tolk or Universal Speech on Windows. I have no clue
> about Mac.

Qt has an API to send arbitrary text to be read out lout in a platform 
independent way:

https://doc.qt.io/qt-5/qtexttospeech.html#say

I was hoping there’s a higher-level feature analogous to live regions but 
for desktop apps which is more standard accross accessibility technologies 
and platforms. If that’s not the case, then I can implement something 
directly using QTextToSpeech::say().

> This should be the first priority, because anything else is unimportant
> until incoming text is automatically read.

I agree.

> Maybe if you can get this working under Linux, someone else can come 
along and duplicate it for Windows.

I hope it doesn’t come to that, but it’s a last resort possibility.

ironcross32 commented:
I'm fairly sure that would use Speech Dispatcher on Linux, which is fine. It would probably even use an acceptable speech system on Mac, but it would use SAPI on Windows which is not good. Sapi 5 is not as responsive as the user's screen reader, and a lot of apps that use it don't offer the end-user the ability to change voices. Further, most of us have voices set up in our screen readers that already work for us. We can change various aspects such as punctuation amount, and add things to the pronunciation dictionary for fantasy names and the like, which the screen reader would not likely pronounce correctly.
@ethindp, can you weigh in on this?

ethindp commented:
One thing you can do is use `QTextToSpeech` on non-Windows platforms, and use Tolk on Windows. Tolk allows you to interface with SAPI, so the TTS needs of the user are fully taken care of with Tolk.

bauermann commented:
> Just an initial review...!

Thank you very much for this detailed review! I just force-pushed my branch with the changes corresponding to the resolved comments. I still have to work on the unresolved ones.

vadi2 commented:
It is _far_ too early to give up on the current cross-platform approach. We've discussed the reasons for this approach previously in https://github.com/Mudlet/Mudlet/issues/3342, let's not get sidetracked here again as well.

vadi2 committed:
Merge branch 'development' into issue-3342-text-window-accessibility

bauermann committed:
TAccessibleTextEdit: Address uninitialized variable warnings from clang-tidy

Example warning:

warning: variable lengthSoFar is not initialized [cppcoreguidelines-init-variables]

int lengthSoFar;
    ^
                = 0

bauermann committed:
TTextEdit: Fix potential memory leak warnings from clang-tidy

Example warning:

warning: Potential memory leak [clang-analyzer-cplusplus.NewDeleteLeaks]
    ⋮
src/TTextEdit.cpp:351:42: note: Memory is allocated
        QAccessible::updateAccessibility(new QAccessibleTextInsertEvent(this,
                                         ^
src/TTextEdit.cpp:355:1: note: Potential memory leak

bauermann commented on src/TTextEdit.cpp:
```diff
@@ -321,6 +330,28 @@ void TTextEdit::showNewLines()
+                                                                        ti->characterCount() - newLines.length(),
+                                                                        newLines));
+    }
 }
```
I *think* I fixed this warning. I wasn't able to run `clang-tidy`, so the next pull request will tell.

bauermann committed:
mudlet: Fix clang-tidy warning about slot_caret_mode() not being const.

bauermann committed:
TAccessibleTextEdit: Fix issues found by CodeFactor

src/TAccessibleTextEdit.cpp#L374
Using deprecated casting style.  Use static_cast<int>(...) instead (readability/casting)

and

src/TAccessibleTextEdit.cpp#L525
Tab found; better to use spaces (whitespace/tab)

vadi2 committed:
Merge branch 'development' into issue-3342-text-window-accessibility

bauermann committed:
TTextEdit: Remove superfluous call to getGraphemeWidth()

I added that call thinking that some locales with many wide characters would
have a wide ‘a’ character as well. I tested with zh_TW and it wasn't the case,
so remove it.

Thanks to Stephen Lyons for noticing the issue.

bauermann committed:
main_window.ui: Change Caret Mode shortcut to ‘Ctrl+Tab’

Thanks to ironcross32 for the suggestion.

vadi2 committed:
Merge branch 'development' into issue-3342-text-window-accessibility

vadi2 committed:
Merge branch 'development' into issue-3342-text-window-accessibility

vadi2 committed:
Merge branch 'development' into issue-3342-text-window-accessibility

vadi2 commented:
See https://github.com/Mudlet/Mudlet/issues/3342#issuecomment-1065913919 for an update. I'll continue this work - will start a new PR as I haven't got the rights to push in here.
